src/main/java/br/unb/cloudissues/http/JavaProjectsRetriever.java

```diff
@@ -1,218 +1,219 @@
+			timeToWaitInMiliSecondsBetweenRequests = time;
+			return this;
+		}
+
+		public Builder withMaxRequestsToWait(int maxRequestsToWait) {
+			this.maxRequestsToWait = maxRequestsToWait;
+			return this;
+		}
+
+		public JavaProjectsRetriever build() {
+			return new JavaProjectsRetriever(this);
+		}
+	}
+
+	public List<Project> retrieve() throws IOException, InterruptedException {
+		Objects.requireNonNull(this.baseUrl);
+
+		count = 1;
+
+		System.out.println("\nretrieving Java projects in " + baseUrl + PROJECTS_SEARCH_URL);
+		return httpRequestJavaProjects();
+
+	}
+
+	private List<Project> httpRequestJavaProjects() throws IOException, InterruptedException {
+		String responseBody = retrieveResponseBodyForURL(buildURL());
+		Map<String, Object> responseMap = Utils.responseToMap(responseBody);
+		@SuppressWarnings("unchecked")
+		Long total = ((Map<String, Double>) responseMap.get("paging")).get("total").longValue();
+
+		List<Project> projects = new ArrayList<>(total.intValue());
+
+		List<SonarProjectComponent> projectsComponents = retrieveSonarProjectComponentsFromResponseMap(
+				responseMap);
+
+		projects.addAll(projectsComponents.stream().map(SonarProjectComponent::toProject)
+				.collect(toList()));
+
+		if (total > DEFAULT_PAGE_SIZE) {
+			projects.addAll(requestsProjectsForMoreThanOnePage(total));
+		}
+
+		return projects;
+	}
+
+	private String buildURL() {
+		HttpUrl.Builder urlBuilder = minimumUrlBuilder();
+		return urlBuilder.build().toString();
+	}
+
+	private HttpUrl.Builder minimumUrlBuilder() {
+		HttpUrl.Builder urlBuilder = Objects
+				.requireNonNull(HttpUrl.parse(String.format("%s%s", baseUrl, PROJECTS_SEARCH_URL)), "Parser of url was returned null")
+				.newBuilder();
+		urlBuilder.addQueryParameter("filter", "languages=java");
+		urlBuilder.addQueryParameter("ps", DEFAULT_PAGE_SIZE.toString());
+		urlBuilder.addQueryParameter("format", "json");
+		return urlBuilder;
+	}
+
+	private String retrieveResponseBodyForURL(String url) throws IOException, InterruptedException {
+		tryNotToFloodSonarWithLotsOfRequests();
+		return doRetrieveResponseBodyForURL(url);
+	}
+
+	private void tryNotToFloodSonarWithLotsOfRequests() throws InterruptedException {
+		if (count % maxRequestsToWait == 0) {
+			System.out.println("\nWaiting for "
+					+ (timeToWaitInMiliSecondsBetweenRequests / 1000.0 + " seconds..."));
+			System.out.println(count + " requests so far.");
+			Thread.sleep(timeToWaitInMiliSecondsBetweenRequests);
+		}
+		count++;
+	}
+
+	private String doRetrieveResponseBodyForURL(String url) throws IOException {
+		Request request = new Request.Builder().url(url).build();
+		Response response = httpClient.newCall(request).execute();
+		return Objects.requireNonNull(response.body(), "The body returned null").string();
+	}
+
+	@SuppressWarnings("unchecked")
+	private List<SonarProjectComponent> retrieveSonarProjectComponentsFromResponseMap(final Map<String, Object> responseMap) {
```

Comment:
### Changes Here
* retrieveSonarProjectComponentsFromResponseMap:
    * Fixing the NullPointerException when the json in the variable called components was returned null;
    * Trying to improve the performance using the stream from java. I think the more advantage is the Lazy Loading operation in the API of Stream
    * I think is the great opportunity to use keyword final to symbolize that variable not change, it's a constant like concepts from value.